wrangler-core/src/main/java/io/cdap/wrangler/utils/ArithmeticOperations.java

```diff
@@ -388,12 +391,55 @@ public static Float min(Float f1, Float f2) {
   }
 
   /**
-   Arithmetic operation - Find the minimum of two columns.
+   * Arithmetic operation - Find the minimum of two columns.
    */
   public static BigDecimal min(BigDecimal bd1, BigDecimal bd2) {
     if (bd1 == null || bd2 == null) {
       return null;
     }
     return bd1.min(bd2);
   }
+
+  /**
+   * Arithmetic operation - calculate average of any number of columns.
+   * Returns Number - BigDecimal if any column is of type BigDecimal, double otherwise.
+   * HALF_EVEN rounding mode chosen for high precision in the long run.
+   */
+  public static Number average(Number... nums) throws DirectiveExecutionException {
+    // Note that this algorithm avoids overflow
+    BigDecimal avg = BigDecimal.ZERO;
+    int t = 1;
+    boolean allNull = true;
+    boolean noBigDecimal = true;
+    for (Number num : nums) {
+      // Check for null input
+      if (num == null) {
+        continue;
+      } else {
+        allNull = false;
+      }
+      // Check for invalid input
+      if (num instanceof Byte) {
+        throw new DirectiveExecutionException("Input cannot be of type 'Byte'.");
+      }
+      // Compute average
+      BigDecimal numer;
```

Comment:
We have just changed the function to perform double operations instead of BigDecimal ones unless there are BigDecimal inputs. As the existence of BigDecimal inputs makes BigDecimal operations unavoidable, should we still try running benchmarks?

Reply:
I think having an idea of the performance would be useful, so yes.

Reply:
Ok, thanks for the help!

Reply:
**Benchmark Summary:** Conducted benchmark tests with 3 CSVs containing 2 columns of randomly generated doubles from 0 to 1, with 100,000 entries in each.

**Results:**
double: 25, 23, 22 sec
decimal: 26, 24, 22 sec

**Notes:** Not all decimal values were written to the output file, as decimal schema requires specification of scale and precision, causing errors in trying to write to output file as scale and precision varies (skipped errors for benchmark). This may have taken away a slight amount of time. In addition, decimal computation required column conversion to decimal type prior to average computation, which may have added a slight amount of time.

**Conclusion:** Times appear to be approximately equal for decimal and double computations at this scale.